crates/ruff_linter/src/rules/flake8_type_checking/rules/runtime_string_union.rs

```diff
@@ -0,0 +1,90 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast as ast;
+use ruff_python_ast::{Expr, Operator};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for the presence of string literals in `X | Y`-style union types.
+///
+/// ## Why is this bad?
+/// [PEP 604] introduced a new syntax for union type annotations based on the
+/// `|` operator.
+///
+/// While Python's type annotations can typically be wrapped in strings to
+/// avoid runtime evaluation, the use of a string member within an `X | Y`-style
+/// union type will cause a runtime error.
+///
+/// Instead, remove the quotes, wrap the _entire_ union in quotes, or use
+/// `from __future__ import annotations` to disable runtime evaluation of
+/// annotations entirely.
+///
+/// ## Example
+/// ```python
+/// var: str | "int"
+/// ```
+///
+/// Use instead:
+/// ```python
+/// var: str | int
+/// ```
+///
+/// ## References
+/// - [PEP 535](https://peps.python.org/pep-0563/)
+/// - [PEP 604](https://peps.python.org/pep-0604/)
+///
+/// [PEP 604]: https://peps.python.org/pep-0604/
+#[violation]
+pub struct RuntimeStringUnion;
+
+impl Violation for RuntimeStringUnion {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Invalid string member in `X | Y`-style union type")
+    }
+}
+
+/// TCH006
+pub(crate) fn runtime_string_union(checker: &mut Checker, expr: &Expr) {
+    if !checker.semantic().in_type_definition() {
+        return;
+    }
+
+    if !checker.semantic().execution_context().is_runtime() {
+        return;
+    }
+
+    // Search for strings within the binary operator.
+    let mut strings = Vec::new();
+    traverse_op(expr, &mut strings);
+
+    for string in strings {
+        checker
+            .diagnostics
+            .push(Diagnostic::new(RuntimeStringUnion, string.range()));
+    }
+}
+
+/// Collect all string members in possibly-nested binary `|` expressions.
+fn traverse_op<'a>(expr: &'a Expr, strings: &mut Vec<&'a Expr>) {
```

Comment:
Does python support indexer types similar to typescript that need traversing?

```
type A = Value[int | string]
```

or is it safe to assume that union types are always top-level?

Edit: I think this is a valid example for when the union is not top level: 

```python
callback: Callable[[str], Awaitable[None | str | "int"]] = on_update
```

or 

```python
# Type checker will infer that all keys in ``z`` are meant to be strings,
# and that all values in ``z`` are meant to be either strings or ints
z: Mapping[str, str | int] = {}
```

Some other examples where I'm not entirely sure if they're valid/we need to handle here

```python
OldS = TypeVar('OldS', int | 'str', str)
```

```python
def inproduct[T: (int | complex, float, complex)](v: Vec[T]) -> T: # Same as Iterable[tuple[T, T]]
    return sum(x*y for x, y in v)
```

Do we need to handle `Annotated`?

Reply:
Yup these are all valid and we should already be handling them correctly outside of the specific context of this rule, since these traversals are encoded in the `Semantic` model (i.e., whether we're in a type definition, and whether it will be evaluated at runtime).

Reply:
Added more tests, thanks :)